crates/wasi-common/src/fdentry.rs

```diff
@@ -1,35 +1,335 @@
+    pub fn new() -> Self {
+        Self {
+            cursor: 0,
+            data: Arc::new(RefCell::new(Vec::new()))
+        }
+    }
+
+    pub fn append(&self, data: &[u8]) {
+        self.data.borrow_mut().extend_from_slice(data);
+    }
+}
+
+impl VirtualFile for InMemoryFile {
+    fn try_clone(&self) -> io::Result<Box<dyn VirtualFile>> {
+        Ok(Box::new(
+            InMemoryFile {
+                cursor: 0,
+                data: Arc::clone(&self.data)
+            }
+        ))
+    }
+
+    fn readlinkat(&self, path: &Path) -> Result<String> {
+        // no symlink support, so always say it's invalid.
+        Err(Error::EINVAL)
+    }
+
+    fn openat(&self, path: &Path, read: bool, write: bool, oflags: u16, fs_flags: u16) -> Result<Box<dyn VirtualFile>> {
+        panic!("oh no");
+    }
+
+    fn write_vectored(&mut self, iovs: &[io::IoSlice]) -> Result<usize> {
+        let mut data = self.data.borrow_mut();
+        let mut cursor = self.cursor;
+        for iov in iovs {
+            for el in iov.iter() {
+                if cursor == data.len() {
+                    data.push(*el);
+                } else {
+                    data[cursor] = *el;
+                }
+                cursor += 1;
+            }
+        }
+        let len = cursor - self.cursor;
+        self.cursor = cursor;
+        Ok(len)
+    }
+
+    fn read_vectored(&mut self, iovs: &mut [io::IoSliceMut]) -> Result<usize> {
+        let data = self.data.borrow();
+        let mut cursor = self.cursor;
+        for iov in iovs {
+            for i in 0..iov.len() {
+                if cursor >= data.len() {
+                    let count = cursor - self.cursor;
+                    self.cursor = cursor;
+                    return Ok(count);
+                }
+                iov[i] = data[cursor];
+                cursor += 1;
+            }
+        }
+
+        let count = cursor - self.cursor;
+        self.cursor = cursor;
+        Ok(count)
+    }
+
+    fn advise(&self, advice: wasi::__wasi_advice_t, offset: wasi::__wasi_filesize_t, len: wasi::__wasi_filesize_t) -> Result<()> {
+        // we'll just ignore advice for now, unless it's totally invalid
+        match advice {
+            wasi::__WASI_ADVICE_DONTNEED |
+            wasi::__WASI_ADVICE_SEQUENTIAL |
+            wasi::__WASI_ADVICE_WILLNEED |
+            wasi::__WASI_ADVICE_NOREUSE |
+            wasi::__WASI_ADVICE_RANDOM |
+            wasi::__WASI_ADVICE_NORMAL => {
+                Ok(())
+            },
+            _ => Err(Error::EINVAL)
+        }
+    }
+
+    fn get_file_type(&self) -> wasi::__wasi_filetype_t {
+        wasi::__WASI_FILETYPE_REGULAR_FILE
+    }
+
+    fn get_rights_base(&self) -> wasi::__wasi_rights_t {
+        wasi::RIGHTS_REGULAR_FILE_BASE
+    }
+
+    fn get_rights_inheriting(&self) -> wasi::__wasi_rights_t {
+        wasi::RIGHTS_REGULAR_FILE_INHERITING
+    }
+}
+
+/// A clonable read/write directory.
+pub struct VirtualDir {
```

Comment:
In the same theme as "Where would `InMemoryFile` live?", I have a virtual directory here as well. I'm satisfied for the moment by being able to say it _could_ support symlinks one day if it was particularly interesting, but again I'm not enthusiastic about suggesting an (incomplete!) virtual filesystem live in wasi-common.

Reply:
I wouldn't worry with symlinks too much at this stage. We still haven't figured out what to do with differences between POSIX vs Windows-based platforms in terms of symlinks which are drastically different!